gui/webdisplay/inc/ROOT/TWebWindowsManager.hxx

```diff
@@ -30,16 +31,23 @@ class THttpWSHandler;
 namespace ROOT {
 namespace Experimental {
 
+class TWebWindowManagerGuard;
+
 class TWebWindowsManager {
 
    friend class TWebWindow;
+   friend class TWebWindowManagerGuard;
 
 private:
    std::unique_ptr<THttpServer> fServer; ///<!  central communication with the all used displays
-   std::string fAddr;                    ///<!   HTTP address of the server
+   std::string fAddr;                    ///<!  HTTP address of the server
+   std::mutex fMutex;                    ///<!  main mutex to protect
+   int fMutexBooked{0};                  ///<!  flag indicating that mutex is booked for some long operation
```

Comment:
Here I want to prevent situation, when mutex locked for very long time - many seconds. 
Therefore I use extra logic which I calling "mutex booking". I lock mutex in the begin, increment extra variable (fMutexBooked) and release lock immediately. At the end of operation variable will be decremented. 
Any other lock can lock such mutex, but will detect that it "booked". And either try it again (as I did by default) or just exit such function with error code.

If I will use only mutex, it will be blocked for long and complex operation and may easily lead to dead locks in the future.

Reply:
If you can have dead locks this means that you 2 distinct but inter-related locks.   In such a case you have two alternatives, one is to merge the locks (i.e. use only one) or make sure the lock is **only** around code that can not (ever) take the other lock.  This can be achieved sometimes by doing first the work/preparation that might take the other lock and then do the small work (eg. assignment) that need the local lock.

For the pattern you use fMutexBooked **must** be an atomic (because it can and will be read and written from multiple threads concurrently.

TWebWindowManagerGuard implements a busy wait for the lock which is likely worse that what you try to avoid.   

And likely the clincher ... this pattern does not prevent a dead lock (unless you ALWAYS use the can't take the look then quit).  [Be cause if thread A take fMutexBooked One then thread B takes fMutexBooked Two and then attempt to take fMutexBooked One [thread 2 starts a busy wait for A to release the critical section] ... then if thread A attempt to take fMutexBooked Two (before release the One) ... then thread A starts an infinite busy wait for fMutexBooked Two to be released ....

Reply:
First of all, I do not have deadlocks at the moment. Main motivation for this was - prevent situation when any other thread requesting for the lock blocked for many seconds (time required to start external web-browser process).

Of course, such approach works only if TWebWindowManagerGuard (or similar technique) used always for the data access - otherwise logic will not work.

My idea - later one could implement more smart technique. Instead of dummy sleep loop of TWebWindowManagerGuard one could put other actions which could be executed in that thread. 

At any moment we could replace this technique with std::recursive_mutex.

Reply:
> First of all, I do not have deadlocks at the moment. Main motivation for this was - prevent situation when any other thread requesting for the lock blocked for many seconds (time required to start external web-browser process).

Then, please use a simpler implementation for now.  I.e. for example a simple mutex and calls to try_lock.  I currently see 'no' advantages to the more complex implementation (that is currently used in the PR).